### PR TITLE
[GISel] Convert zext nneg to sext if it is cheaper

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch converts `zext nneg` to `sext` on RISCV to use free sext.

It is an alternative to https://github.com/llvm/llvm-project/pull/93842.

### CODE REVIEW

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 1]
```diff
@@ -7400,3 +7400,23 @@ bool CombinerHelper::matchZextOfTrunc(const MachineOperand &MO,
 
   return false;
 }
+
+bool CombinerHelper::matchNonNegZext(const MachineOperand &MO,
+                                     BuildFnTy &MatchInfo) {
+  GZext *Zext = cast<GZext>(getDefIgnoringCopies(MO.getReg(), MRI));
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:7406]: Can this just to get def? It shouldn't have to worry about intervening copies (I think we have too many uses of getDefIgnoringCopies)

File: llvm/include/llvm/Target/GlobalISel/Combine.td [hunk 2]
```diff
@@ -1793,7 +1800,7 @@ def all_combines : GICombineGroup<[integer_reassoc_combines, trivial_combines,
     sub_add_reg, select_to_minmax, redundant_binop_in_equality,
     fsub_to_fneg, commute_constant_to_rhs, match_ands, match_ors,
     combine_concat_vector, double_icmp_zero_and_or_combine, match_addos,
-    sext_trunc, zext_trunc, combine_shuffle_concat,
+    sext_trunc, zext_trunc, nneg_zext, combine_shuffle_concat,
```
Comment (arsenm) [inline llvm/include/llvm/Target/GlobalISel/Combine.td:1803]: Can we start organizing this into some kind of prefer-signed grouping? Probably should also have a reasonable-but-not-all set that excludes bits like this that are more target dependent

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 3]
```diff
@@ -7407,3 +7407,23 @@ bool CombinerHelper::matchZextOfTrunc(const MachineOperand &MO,
...
+  Register Dst = Zext->getReg(0);
+  Register Src = Zext->getSrcReg();
+
+  LLT DstTy = MRI.getType(Dst);
+  LLT SrcTy = MRI.getType(Src);
+  const auto &TLI = getTargetLowering();
+
+  // Convert zext nneg to sext if sext is the preferred form for the target.
+  if (TLI.isSExtCheaperThanZExt(getMVTForLLT(SrcTy), getMVTForLLT(DstTy))) {
+    MatchInfo = [=](MachineIRBuilder &B) { B.buildSExt(Dst, Src); };
```
Comment (tschuett) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:7424]:
```
if (!isLegalOrBeforeLegalizer({TargetOpcode::G_SEXT, {DstTy, SrcTy}}))
  return false;
```
We have to check whether the sext is legal or we are before the legalizer..

**Comment by tschuett** [comment]
Nonetheless, nice first combine!

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp (hunk 3 above)
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:7424]: It is only cheaper. You had to use MVT types to ask for cheapness. It is a DAGish call. You still have to query the GlobalIsel legalizer for legalness.

File: llvm/include/llvm/Target/GlobalISel/Combine.td [hunk 4]
```diff
@@ -1780,6 +1780,8 @@ def fma_combines : GICombineGroup<[combine_fadd_fmul_to_fmad_or_fma,
 def constant_fold_binops : GICombineGroup<[constant_fold_binop,
                                            constant_fold_fp_binop]>;
 
+def prefer_sign_combines : GICombineGroup<[nneg_zext]>;
+
 def all_combines : GICombineGroup<[integer_reassoc_combines, trivial_combines,
```
Comment (arsenm) [inline llvm/include/llvm/Target/GlobalISel/Combine.td:1785]: we probably should have a "mostly" target independent group, but that's another change

File: llvm/test/CodeGen/RISCV/GlobalISel/alu-roundtrip-rv64.ll [hunk 5]
```diff
@@ -101,3 +101,13 @@ entry:
...
+
+define i64 @zext_nneg_i32_i64(i32 %a) {
+; RV64IM-LABEL: zext_nneg_i32_i64:
+; RV64IM:       # %bb.0: # %entry
+; RV64IM-NEXT:    sext.w a0, a0
+; RV64IM-NEXT:    ret
+entry:
+  %b = zext nneg i32 %a to i64
+  ret i64 %b
+}
```
Comment (arsenm) [inline llvm/test/CodeGen/RISCV/GlobalISel/alu-roundtrip-rv64.ll:113]: Vector case

**Review by arsenm — COMMENTED** [review]
lgtm with vector test

File: llvm/test/CodeGen/RISCV/GlobalISel/alu-roundtrip-rv64.ll (hunk 5 above)
Comment (dtcxzyw) [reply llvm/test/CodeGen/RISCV/GlobalISel/alu-roundtrip-rv64.ll:113]: Sorry, I cannot provide vector tests as both RISC-V and LoongArch only handle scalar i32 -> i64.
https://github.com/llvm/llvm-project/blob/a65771fce4a2f25f16d4b3918ad6a11370637f7b/llvm/lib/Target/RISCV/RISCVISelLowering.cpp#L1881-L1883
https://github.com/llvm/llvm-project/blob/a65771fce4a2f25f16d4b3918ad6a11370637f7b/llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp#L4982-L4985

**Comment by dtcxzyw** [comment]
Any questions? Can anyone give me an explicit approval?

**Review by arsenm — APPROVED** [review]

**Pull request merged**